File: julia/test/cate/test_neural_dml.jl
#=
Tests for Neural Double Machine Learning with Cross-Fitting

Session 155: Julia Neural CATE Parity

Test Layers:
- Layer 1: Known-Answer - Cross-fitting correctness
- Layer 2: Adversarial - Edge cases and error handling
=#

using Test
using CausalEstimators
using Random
using Statistics

# Include shared DGP file
include("dgp_cate.jl")


# =============================================================================
# Layer 1: Known-Answer Tests
# =============================================================================

@testset "Neural DML Known-Answer" begin

    @testset "Constant effect DGP" begin
        data = dgp_constant_effect(n=1000, true_ate=2.0, seed=42)
        problem = CATEProblem(data.Y, data.treatment, data.X, (alpha=0.05,))

        solution = solve(problem, NeuralDoubleMachineLearning(n_folds=5, degree=2))

        @test solution.retcode == :Success
        @test solution.method == :neural_dml
        @test length(solution.cate) == data.n

        # ATE should be close to true
        @test abs(solution.ate - data.true_ate) < 0.5

        # CI should cover true ATE
        @test solution.ci_lower < data.true_ate < solution.ci_upper
    end

    @testset "Linear heterogeneity DGP" begin
        data = dgp_linear_heterogeneity(n=1000, base_effect=2.0, het_coef=1.0, seed=42)
        problem = CATEProblem(data.Y, data.treatment, data.X, (alpha=0.05,))

        solution = solve(problem, NeuralDoubleMachineLearning(n_folds=5, degree=2))

        @test abs(solution.ate - data.true_ate) < 0.5

        # CATE should correlate with truth
        corr = cor(solution.cate, data.true_cate)
        @test corr > 0.5
    end

    @testset "Strong confounding DGP" begin
        # DML with cross-fitting should handle confounding
        data = dgp_strong_confounding(n=1000, seed=42)
        problem = CATEProblem(data.Y, data.treatment, data.X, (alpha=0.05,))

        solution = solve(problem, NeuralDoubleMachineLearning(n_folds=5, degree=2))

        @test solution.retcode == :Success
        # Allow larger tolerance for confounded data
        @test abs(solution.ate - data.true_ate) < 1.0
    end

    @testset "Different fold counts" begin
        data = dgp_constant_effect(n=500, true_ate=2.0, seed=42)
        problem = CATEProblem(data.Y, data.treatment, data.X, (alpha=0.05,))

        # Test different fold counts
        for n_folds in [2, 3, 5, 10]
            solution = solve(problem, NeuralDoubleMachineLearning(n_folds=n_folds))
            @test solution.retcode == :Success
            @test abs(solution.ate - data.true_ate) < 1.0
        end
    end

end


# =============================================================================
# Layer 2: Adversarial Tests
# =============================================================================

@testset "Neural DML Adversarial" begin

    @testset "Invalid n_folds parameter" begin
        @test_throws ArgumentError NeuralDoubleMachineLearning(n_folds=1)
        @test_throws ArgumentError NeuralDoubleMachineLearning(n_folds=0)
    end

    @testset "Invalid degree parameter" begin
        @test_throws ArgumentError NeuralDoubleMachineLearning(degree=0)
    end

    @testset "Invalid lambda parameter" begin
        @test_throws ArgumentError NeuralDoubleMachineLearning(lambda=-0.1)
    end

    @testset "Insufficient samples for cross-fitting" begin
        Random.seed!(42)
        n = 15  # Too small for 5-fold cross-fitting (3 per fold)
        X = randn(n, 3)
        T = rand(n) .< 0.5
        Y = X[:, 1] .+ 2.0 .* T .+ randn(n) * 0.5

        problem = CATEProblem(Y, T, X, (alpha=0.05,))

        # Should throw error for insufficient fold size
        @test_throws ArgumentError solve(problem, NeuralDoubleMachineLearning(n_folds=5))
    end

    @testset "High-dimensional covariates" begin
        data = dgp_high_dimensional(n=500, p=30, seed=42)
        problem = CATEProblem(data.Y, data.treatment, data.X, (alpha=0.05,))

        # Use degree=1 and higher lambda for high-dim
        solution = solve(problem, NeuralDoubleMachineLearning(
            n_folds=5, degree=1, lambda=1.0
        ))

        @test solution.retcode == :Success
        @test isfinite(solution.ate)
        @test isfinite(solution.se)
    end

    @testset "Imbalanced treatment" begin
        data = dgp_imbalanced_treatment(n=500, true_ate=2.0, treatment_prob=0.15, seed=42)
        problem = CATEProblem(data.Y, data.treatment, data.X, (alpha=0.05,))

        solution = solve(problem, NeuralDoubleMachineLearning(n_folds=5, degree=2))

        @test solution.retcode == :Success
        @test abs(solution.ate - data.true_ate) < 1.5
    end

end


# =============================================================================
# Cross-Fitting Verification
# =============================================================================

@testset "Neural DML Cross-Fitting Properties" begin

    @testset "More folds → more variance but less bias" begin
        data = dgp_constant_effect(n=500, true_ate=2.0, seed=42)
        problem = CATEProblem(data.Y, data.treatment, data.X, (alpha=0.05,))

        ates = Float64[]
        for seed in 1:20
            Random.seed!(seed)
            # Use same problem, different random splits
            sol = solve(problem, NeuralDoubleMachineLearning(n_folds=5))
            push!(ates, sol.ate)
        end

        # Mean should be close to true ATE
        @test abs(mean(ates) - 2.0) < 0.3

        # Some variance across random splits is expected
        @test std(ates) > 0.01
        @test std(ates) < 0.5
    end

end


# =============================================================================
# Comparison Tests
# =============================================================================

@testset "Neural DML vs Standard DML" begin

    @testset "Both recover ATE on linear DGP" begin
        data = dgp_linear_heterogeneity(n=500, seed=42)
        problem = CATEProblem(data.Y, data.treatment, data.X, (alpha=0.05,))

        sol_dml = solve(problem, DoubleMachineLearning(n_folds=5))
        sol_neural = solve(problem, NeuralDoubleMachineLearning(n_folds=5, degree=2))

        # Both should be close to true ATE
        @test abs(sol_dml.ate - data.true_ate) < 0.5
        @test abs(sol_neural.ate - data.true_ate) < 0.5

        # Both should have valid CIs
        @test sol_dml.se > 0
        @test sol_neural.se > 0
    end

    @testset "Neural DML captures nonlinear heterogeneity better" begin
        data = dgp_nonlinear_heterogeneity(n=1000, seed=42)
        problem = CATEProblem(data.Y, data.treatment, data.X, (alpha=0.05,))

        sol_dml = solve(problem, DoubleMachineLearning(n_folds=5))
        sol_neural = solve(problem, NeuralDoubleMachineLearning(n_folds=5, degree=2))

        # Neural should have at least comparable CATE correlation
        # (polynomial features help with nonlinear patterns)
        corr_dml = cor(sol_dml.cate, data.true_cate)
        corr_neural = cor(sol_neural.cate, data.true_cate)

        # Both should capture some heterogeneity
        @test corr_dml > 0.2
        @test corr_neural > 0.2
    end

end

File: tests/test_cate/test_neural_dml.py
"""Tests for neural Double Machine Learning.

Test layers:
- Layer 1: Known-answer tests (ATE recovery, cross-fitting validation)
- Layer 2: Adversarial tests (sample size, fold counts)
- Layer 3: Monte Carlo validation (bias, coverage) @slow
"""

import numpy as np
import pytest

from causal_inference.cate.neural_dml import neural_double_ml

from .conftest import generate_cate_dgp


# ============================================================================
# Layer 1: Known-Answer Tests
# ============================================================================


class TestNeuralDMLKnownAnswer:
    """Known-answer tests for neural Double ML."""

    def test_constant_effect_ate_recovery(self, constant_effect_data):
        """Neural DML should recover ATE within tolerance."""
        Y, T, X, _ = constant_effect_data
        result = neural_double_ml(Y, T, X, n_folds=3)

        assert abs(result["ate"] - 2.0) < 0.5, f"ATE {result['ate']} not close to 2.0"

    def test_cate_shape(self, constant_effect_data):
        """CATE should have correct shape."""
        Y, T, X, _ = constant_effect_data
        result = neural_double_ml(Y, T, X, n_folds=3)

        assert result["cate"].shape == (len(Y),)

    def test_ci_validity(self, constant_effect_data):
        """CI should be valid (lower < upper)."""
        Y, T, X, _ = constant_effect_data
        result = neural_double_ml(Y, T, X, n_folds=3)

        assert result["ci_lower"] < result["ci_upper"]

    def test_se_positive(self, constant_effect_data):
        """SE should be positive."""
        Y, T, X, _ = constant_effect_data
        result = neural_double_ml(Y, T, X, n_folds=3)

        assert result["ate_se"] > 0

    def test_method_name(self, constant_effect_data):
        """Method name should be correct."""
        Y, T, X, _ = constant_effect_data
        result = neural_double_ml(Y, T, X, n_folds=3)

        assert result["method"] == "neural_double_ml"

    def test_heterogeneous_effect_correlation(self, linear_heterogeneous_data):
        """Neural DML should capture heterogeneity."""
        Y, T, X, true_cate = linear_heterogeneous_data
        result = neural_double_ml(Y, T, X, n_folds=3)

        correlation = np.corrcoef(result["cate"], true_cate)[0, 1]
        assert correlation > 0.2, f"Correlation {correlation} too low"


# ============================================================================
# Layer 2: Adversarial Tests
# ============================================================================


class TestNeuralDMLAdversarial:
    """Adversarial tests for neural Double ML."""

    def test_insufficient_samples_raises(self):
        """Should raise error when too few samples for cross-fitting."""
        Y, T, X, _ = generate_cate_dgp(n=20, p=2, seed=42)

        with pytest.raises(ValueError, match="Insufficient samples"):
            neural_double_ml(Y, T, X, n_folds=5)

    def test_two_fold_minimum(self, constant_effect_data):
        """Should work with minimum 2 folds."""
        Y, T, X, _ = constant_effect_data
        result = neural_double_ml(Y, T, X, n_folds=2)

        assert isinstance(result["ate"], float)

    def test_high_dimensional(self, high_dimensional_data):
        """Should handle high-dimensional covariates."""
        Y, T, X, _ = high_dimensional_data
        result = neural_double_ml(Y, T, X, n_folds=3)

        assert not np.isnan(result["ate"])

    def test_single_covariate(self, single_covariate_data):
        """Should handle single covariate."""
        Y, T, X, _ = single_covariate_data
        result = neural_double_ml(Y, T, X.ravel(), n_folds=3)

        assert isinstance(result["ate"], float)

    def test_empty_treatment_group_raises(self):
        """Should raise error when treatment group is empty."""
        n = 100
        Y = np.random.randn(n)
        T = np.zeros(n)  # All control
        X = np.random.randn(n, 2)

        with pytest.raises(ValueError, match="treatment"):
            neural_double_ml(Y, T, X, n_folds=3)

    def test_custom_hidden_layers(self, constant_effect_data):
        """Should accept custom hidden layer configuration."""
        Y, T, X, _ = constant_effect_data
        result = neural_double_ml(Y, T, X, n_folds=3, hidden_layers=(50, 25))

        assert isinstance(result["ate"], float)

    def test_custom_alpha(self, constant_effect_data):
        """Should use custom alpha for CI."""
        Y, T, X, _ = constant_effect_data
        result_95 = neural_double_ml(Y, T, X, n_folds=3, alpha=0.05)
        result_99 = neural_double_ml(Y, T, X, n_folds=3, alpha=0.01)

        # 99% CI should be wider than 95% CI
        width_95 = result_95["ci_upper"] - result_95["ci_lower"]
        width_99 = result_99["ci_upper"] - result_99["ci_lower"]
        assert width_99 > width_95

    def test_imbalanced_treatment(self):
        """Should handle imbalanced treatment (20% treated)."""
        Y, T, X, _ = generate_cate_dgp(
            n=500,
            p=2,
            effect_type="constant",
            true_ate=2.0,
            treatment_prob=0.2,  # 20% is more reasonable for cross-fitting
            seed=42,
        )
        result = neural_double_ml(Y, T, X, n_folds=3)

        # Should still get reasonable estimate (looser tolerance for imbalanced)
        assert abs(result["ate"] - 2.0) < 1.5


# ============================================================================
# Layer 3: Monte Carlo Tests
# ============================================================================


class TestNeuralDMLMonteCarlo:
    """Monte Carlo validation for neural Double ML."""

    @pytest.mark.slow
    def test_neural_dml_bias(self):
        """Neural DML should have bias < 0.10 over Monte Carlo runs."""
        n_runs = 100
        estimates = []

        for seed in range(n_runs):
            Y, T, X, _ = generate_cate_dgp(n=300, p=2, true_ate=2.0, seed=seed)
            result = neural_double_ml(Y, T, X, n_folds=3)
            estimates.append(result["ate"])

        bias = abs(np.mean(estimates) - 2.0)
        assert bias < 0.15, f"Neural DML bias {bias:.3f} exceeds threshold"

    @pytest.mark.slow
    def test_neural_dml_coverage(self):
        """Neural DML should have 85-99% coverage."""
        n_runs = 100
        covers = []

        for seed in range(n_runs):
            Y, T, X, _ = generate_cate_dgp(n=300, p=2, true_ate=2.0, seed=seed)
            result = neural_double_ml(Y, T, X, n_folds=3)
            covers.append(result["ci_lower"] < 2.0 < result["ci_upper"])

        coverage = np.mean(covers)
        assert 0.80 < coverage < 0.99, f"Coverage {coverage:.2%} outside 80-99%"

    @pytest.mark.slow
    def test_cross_fitting_reduces_bias(self):
        """Cross-fitting should help reduce bias compared to naive approach."""
        # This test verifies that cross-fitting provides benefit
        n_runs = 50
        estimates_3fold = []
        estimates_5fold = []

        for seed in range(n_runs):
            Y, T, X, _ = generate_cate_dgp(n=400, p=2, true_ate=2.0, seed=seed)
            result_3 = neural_double_ml(Y, T, X, n_folds=3)
            result_5 = neural_double_ml(Y, T, X, n_folds=5)

            estimates_3fold.append(result_3["ate"])
            estimates_5fold.append(result_5["ate"])

        # Both should have reasonable bias
        bias_3 = abs(np.mean(estimates_3fold) - 2.0)
        bias_5 = abs(np.mean(estimates_5fold) - 2.0)

        assert bias_3 < 0.20, f"3-fold bias {bias_3:.3f} too high"
        assert bias_5 < 0.20, f"5-fold bias {bias_5:.3f} too high"

    @pytest.mark.slow
    def test_heterogeneity_recovery(self):
        """Neural DML should recover heterogeneous effects."""
        correlations = []

        for seed in range(50):
            Y, T, X, true_cate = generate_cate_dgp(
                n=500, p=2, effect_type="linear", true_ate=2.0, seed=seed
            )
            result = neural_double_ml(Y, T, X, n_folds=3)
            correlations.append(np.corrcoef(result["cate"], true_cate)[0, 1])

        mean_corr = np.mean(correlations)
        assert mean_corr > 0.2, f"Mean CATE correlation {mean_corr:.2f} too low"
